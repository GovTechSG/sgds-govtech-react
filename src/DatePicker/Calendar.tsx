import * as React from 'react';
interface CalendarProps extends React.HTMLAttributes<HTMLTableElement> {
  selectedDate: Date[];
  displayDate: Date;
  minDate?: string;
  maxDate?: string;
  changeDate: (date: Date) => void;
  dayLabels?: string[];
  mode: 'single' | 'range';
}
const CELL_PADDING = '5px';
const DAY_LABELS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

const setTimeToNoon = (date: Date) => {
  const newDate = new Date(date);
  newDate.setHours(12);
  newDate.setMinutes(0);
  newDate.setSeconds(0);
  newDate.setMilliseconds(0);
  return newDate;
};
const generateIncrementDays = (start: Date, end: Date) => {
  let arr = [];
  for (let dt = start; dt <= end; dt.setDate(dt.getDate() + 1)) {
    arr.push(new Date(dt));
  }
  return arr;
};
export const Calendar = React.forwardRef<HTMLTableElement, CalendarProps>(
  ({ dayLabels = DAY_LABELS, ...props }, ref) => {
    const handleClick = (e: React.MouseEvent<HTMLTableCellElement>) => {
      const day = e.currentTarget.getAttribute('data-day')!;
      const displayDateClone = new Date(props.displayDate);
      const newSelectedDate = setTimeToNoon(displayDateClone);
      newSelectedDate.setDate(parseInt(day));
      props.changeDate(newSelectedDate);
    };
    const currentDate = setTimeToNoon(new Date());
    const selectedDates = props.selectedDate.map((d) => setTimeToNoon(d));
    const rangeSelectedDates = generateIncrementDays(
      new Date(selectedDates[0]),
      new Date(selectedDates[1])
    );
    const minimumDate = props.minDate
      ? setTimeToNoon(new Date(props.minDate))
      : null;
    const maximumDate = props.maxDate
      ? setTimeToNoon(new Date(props.maxDate))
      : null;
    const year = props.displayDate.getFullYear();
    const month = props.displayDate.getMonth();
    const firstDay = new Date(year, month, 1);
    const startingDay = firstDay.getDay();
    let monthLength = daysInMonth[month];
    if (month == 1) {
      if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) {
        monthLength = 29;
      }
    }

    const weeks = [];
    let day = 1;
    for (let i = 0; i < 9; i++) {
      const week = [];
      for (let j = 0; j <= 6; j++) {
        if (day <= monthLength && (i > 0 || j >= startingDay)) {
          let className = undefined;
          const date = new Date(year, month, day, 12, 0, 0, 0).toISOString();
          const beforeMinDate =
            minimumDate &&
            Date.parse(date) < Date.parse(minimumDate.toISOString());
          const afterMinDate =
            maximumDate &&
            Date.parse(date) > Date.parse(maximumDate.toISOString());

          let clickHandler: React.MouseEventHandler | undefined = handleClick;
          const style = {
            cursor: 'pointer',
            padding: CELL_PADDING,
            borderRadius: 50,
          };

          if (beforeMinDate || afterMinDate) {
            className = 'text-muted';
            clickHandler = undefined;
            style.cursor = 'default';
          }

          if (Date.parse(date) === Date.parse(currentDate.toISOString())) {
            //if selected Date is not current Date
            className = 'text-primary';
          }
          if (selectedDates.length > 0) {
            rangeSelectedDates.forEach((d) => {
              if (Date.parse(date) === Date.parse(d.toISOString())) {
                className = 'bg-primary';
              }
            });
            if (
              Date.parse(date) === Date.parse(selectedDates[0]!.toISOString())
            ) {
              className = 'bg-primary';
            }
          }

          week.push(
            <td
              key={j}
              data-day={day}
              onClick={clickHandler}
              style={style}
              className={className}
            >
              {day}
            </td>
          );
          day++;
        } else {
          week.push(<td key={j} />);
        }
      }

      weeks.push(<tr key={i}>{week}</tr>);
      if (day > monthLength) {
        break;
      }
    }

    return (
      <table className="text-center" ref={ref}>
        <thead>
          <tr>
            {dayLabels.map((label: string, index: number) => {
              return (
                <td
                  key={index}
                  className="text-muted"
                  style={{ padding: CELL_PADDING }}
                >
                  <small>{label}</small>
                </td>
              );
            })}
          </tr>
        </thead>
        <tbody>{weeks}</tbody>
      </table>
    );
  }
);

export default Calendar;
