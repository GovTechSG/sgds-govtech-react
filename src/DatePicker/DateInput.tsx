import useMergedRefs from '@restart/hooks/useMergedRefs';
import { useDropdownToggle } from '@restart/ui/DropdownToggle';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import * as React from 'react';
import FormControl, { FormControlProps } from '../Form/FormControl';
import { BsPrefixRefForwardingComponent } from '../utils/helpers';
import useWrappedRefWithWarning from '../utils/useWrappedRefWithWarning';

export interface DateInputProps extends Omit<FormControlProps, 'type'> {
  as?: React.ElementType;
}

type DateInputPropsComponent = BsPrefixRefForwardingComponent<
  'input',
  DateInputProps
>;

const propTypes = {
  /**
   * An html id attribute, necessary for assistive technologies, such as screen readers.
   * @type {string|number}
   */
  id: PropTypes.string,

  as: PropTypes.elementType,
};

const DateInput: DateInputPropsComponent = React.forwardRef(
  (
    {
      className,
      // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
      as: Component = FormControl,
      ...props
    }: DateInputProps,
    ref
  ) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
    const [toggleProps] = useDropdownToggle();
    toggleProps.ref = useMergedRefs(
      toggleProps.ref,
      useWrappedRefWithWarning(ref, 'DropdownToggle')
    );
    // Assign the ref element of dropdown toggle to HTMLInputElement (FormControl)
    React.useEffect(() => {
      toggleProps.ref(inputRef.current);
    }, []);

    return (
      <Component className={classNames(className)} ref={inputRef} {...props} />
    );
  }
);

DateInput.displayName = 'DateInput';
DateInput.propTypes = propTypes;

export default DateInput;
