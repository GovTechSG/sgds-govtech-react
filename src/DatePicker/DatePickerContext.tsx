import * as React from 'react';

export type CalendarView = 'day' | 'month' | 'year';
export interface DatePickerContextProps {
  view: CalendarView;
  setView: React.Dispatch<React.SetStateAction<CalendarView>>;
  focusedDateIndex: number;
  setFocusedDateIndex: React.Dispatch<React.SetStateAction<number>>;
  focusedMonthIndex: number;
  setFocusedMonthIndex: React.Dispatch<React.SetStateAction<number>>;
  focusedYearIndex: number;
  setFocusedYearIndex: React.Dispatch<React.SetStateAction<number>>;
  yearPositionIndex: number; // to keep track of the position of the selected year, without this, year range position will not be consistent after selection
  setYearPositionIndex: React.Dispatch<React.SetStateAction<number>>;
}
const context = React.createContext<DatePickerContextProps>({
  view: 'day',
  setView: () => {},
  focusedDateIndex: 0,
  setFocusedDateIndex: () => {},
  focusedMonthIndex: 0,
  setFocusedMonthIndex: () => {},
  focusedYearIndex: 0,
  setFocusedYearIndex: () => {},
  yearPositionIndex: 0,
  setYearPositionIndex: () => {},
});

context.displayName = 'DatePickerContext';

export default context;
