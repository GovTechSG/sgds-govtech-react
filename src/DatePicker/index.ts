export { default as DatePicker } from './DatePicker';
export type { DatePickerProps, RangeSelectionValue } from './DatePicker';