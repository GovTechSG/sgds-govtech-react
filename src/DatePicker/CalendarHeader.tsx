import * as React from 'react';
import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
interface CalendarHeaderProps {
  displayDate: Date;
  onChange: (date: Date) => void;
  monthLabels: string[];
}
const MONTH_LABELS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

const CalendarHeader: React.FC<CalendarHeaderProps> = ({
  monthLabels = MONTH_LABELS,
  ...props
}) => {
  const { view, setView } = useContext(DatePickerContext);

  const handleClickPrevious = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);
    if (view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 1);
      props.onChange(newDisplayDate);
    }

    if (view === 'year') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 10);
      props.onChange(newDisplayDate);
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() - 1);
    props.onChange(newDisplayDate);
  };

  const handleClickNext = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);

    if (view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 1);
      props.onChange(newDisplayDate);
    }

    if (view === 'year') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 10);
      props.onChange(newDisplayDate);
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() + 1);
    props.onChange(newDisplayDate);
  };

  const changeView = () => {
    switch (view) {
      case 'day':
        setView('month');
        break;
      case 'month':
        setView('year');
        break;
      case 'year':
        break;
    }
  };
  const renderHeader = () => {
    if (view === 'month') return `${props.displayDate.getFullYear()}`;
    if (view === 'year')
      return `${props.displayDate.getFullYear() - 5} - ${
        props.displayDate.getFullYear() + 6
      }`;
    return `${
      monthLabels[props.displayDate.getMonth()]
    } ${props.displayDate.getFullYear()}`;
  };
  return (
    <div className="text-center d-flex justify-content-around">
      <div
        className="text-muted pull-left"
        onClick={handleClickPrevious}
        style={{ cursor: 'pointer' }}
      >
        <i className="bi bi-chevron-left"></i>
      </div>
      <button onClick={changeView} disabled={view === 'year'}>
        {renderHeader()}
      </button>
      <div
        className="text-muted pull-right"
        onClick={handleClickNext}
        style={{ cursor: 'pointer' }}
      >
        <i className="bi bi-chevron-right"></i>
      </div>
    </div>
  );
};
export default CalendarHeader;
