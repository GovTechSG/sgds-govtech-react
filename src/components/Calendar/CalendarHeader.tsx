import Button from '@restart/ui/esm/Button';
import * as React from 'react';
import Dropdown from '../Dropdown/Dropdown';
import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
interface CalendarHeaderProps {
  displayDate: Date;
  minDate?: string;
  maxDate?: string;
  onChange: (date: Date) => void;
  monthLabels: string[];
  previousButtonElement: string | JSX.Element;
  nextButtonElement: string | JSX.Element;
  years?: number[];
}
const MONTH_LABELS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];
const YEARS = (): number[] => {
  const currentYear = new Date().getFullYear();
  const yearsArray = [];
  const start = currentYear - 50;
  const end = currentYear + 50;
  for (let i = start; i < end; i++) {
    yearsArray.push(i);
  }
  return yearsArray;
};
const CalendarHeader: React.FC<CalendarHeaderProps> = ({
  monthLabels = MONTH_LABELS,
  previousButtonElement = '<',
  nextButtonElement = '>',
  years = YEARS(),
  ...props
}) => {
  const { view, setView } = useContext(DatePickerContext);
  const [year, setYear] = React.useState(props.displayDate.getFullYear());
  const displayingMinMonth = () => {
    if (!props.minDate) return false;

    const displayDate = new Date(props.displayDate);
    const minimumDate = new Date(props.minDate);
    return (
      minimumDate.getFullYear() == displayDate.getFullYear() &&
      minimumDate.getMonth() == displayDate.getMonth()
    );
  };

  const displayingMaxMonth = () => {
    if (!props.maxDate) return false;
    const displayDate = new Date(props.displayDate);
    const maximumDate = new Date(props.maxDate);
    return (
      maximumDate.getFullYear() == displayDate.getFullYear() &&
      maximumDate.getMonth() == displayDate.getMonth()
    );
  };

  const handleClickPrevious = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);
    if(view === 'month'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 1)
      props.onChange(newDisplayDate)
    }

    if(view === 'year'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 10)
      props.onChange(newDisplayDate)
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() - 1);
    props.onChange(newDisplayDate);
  };

  const handleClickNext = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);

    if(view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 1)
      props.onChange(newDisplayDate)
    }

    if(view === 'year'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 10)
      props.onChange(newDisplayDate)
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() + 1);
    props.onChange(newDisplayDate);
  };
  const handleSelectYear = (key: string | null) => {
    const newDate = props.displayDate;
    newDate.setFullYear(parseInt(key!));
    props.onChange(newDate);
    setYear(parseInt(key!));
  };
  const [header, setHeader] = React.useState(
    `${monthLabels[props.displayDate.getMonth()]} ${year}`
  );

  const changeView = () => {
    switch (view) {
      case 'day':
        setView('month');
        setHeader(`${year}`);
        break;
      case 'month':
        setView('year');
        break;
      case 'year':
        break;
    }
  };
  const renderHeader = () => {
    if (view === 'month') return `${props.displayDate.getFullYear()}`;
    if (view === 'year') return `${props.displayDate.getFullYear() - 5} - ${props.displayDate.getFullYear() + 6}`;
    return `${monthLabels[props.displayDate.getMonth()]} ${props.displayDate.getFullYear()}`;
  };
  return (
    <div className="text-center d-flex justify-content-around">
      <div
        className="text-muted pull-left"
        onClick={handleClickPrevious}
        style={{ cursor: 'pointer' }}
      >
        {displayingMinMonth() ? null : previousButtonElement}
      </div>
      <Button onClick={changeView} disabled={view === 'year'}>
        {renderHeader()}
      </Button>
      {/*    <Dropdown onSelect={handleSelectYear}>
        <Dropdown.Toggle variant="white" >
          {year}
        </Dropdown.Toggle>
        <Dropdown.Menu >
          {years.map((yr: number) => (
            <Dropdown.Item
            key={yr}
            active={yr===year}
              eventKey={yr}
            >
              {yr}
            </Dropdown.Item>
          ))}
        </Dropdown.Menu>
      </Dropdown> */}
      <div
        className="text-muted pull-right"
        onClick={handleClickNext}
        style={{ cursor: 'pointer' }}
      >
        {displayingMaxMonth() ? null : nextButtonElement}
      </div>
    </div>
  );
};
export default CalendarHeader;
