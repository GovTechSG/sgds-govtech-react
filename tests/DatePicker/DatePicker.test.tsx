import * as React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import {
  DatePicker,
  makeInputValueString,
} from '../../src/DatePicker/DatePicker';
import { MONTH_LABELS } from '../../src/DatePicker/CalendarHeader';

describe('makeInputValueString', () => {
  it('takes in undefined value, returns empty string', () => {
    const arrange = makeInputValueString(undefined, 'DD/MM/YYYY')
    expect(arrange).toEqual('')
  })
  it('takes in Date value, returns string Date in the given format', () => {
    const arrange = makeInputValueString(new Date("2020-01-01"), 'DD/MM/YYYY')
    expect(arrange).toEqual('01/01/2020')
    const arrange1 = makeInputValueString(new Date("2020-01-01"), 'YYYY/MM/DD')
    expect(arrange1).toEqual('2020/01/01')
    const arrange2 = makeInputValueString(new Date("2020-09-23"), 'MM/DD/YYYY')
    expect(arrange2).toEqual('09/23/2020')
  })
})
describe('DatePicker', () => {
  it('has the default html structure', async () => {
    const { container, asFragment, getByText } = render(<DatePicker />);
    expect(asFragment()).toMatchSnapshot();
    expect(container.querySelector('input')).toBeInTheDocument();
    expect(container.querySelector('.input-group')).toHaveAttribute(
      'variant',
      'has-icon'
    );
    expect(container.querySelector('button')).toBeInTheDocument();
    expect(container.querySelector('i.bi-x')).toBeInTheDocument();
    expect(
      container.querySelector('i.form-control-icon.bi-calendar')
    ).toBeInTheDocument();
    expect(container.querySelector('.popover')).not.toBeInTheDocument();

    fireEvent.focus(container.querySelector('input')!);
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument();
      const today = new Date();
      expect(
        getByText(`${MONTH_LABELS[today.getMonth()]} ${today.getFullYear()}`)
      ).toBeInTheDocument();
    });
  });

  it('initialValue passed in, should reflect on input and calendar', async () => {
    const initialValue = new Date();
    const day = initialValue.getDate();
    const initialValueStr = makeInputValueString(initialValue, 'DD/MM/YYYY');
    const { container, getByText } = render(
      <DatePicker initialValue={initialValue} />
    );
    expect(container.querySelector('input')?.value).toEqual(initialValueStr);

    fireEvent.focus(container.querySelector('input')!);
    await waitFor(() =>
      expect(getByText(day).classList).toContain('bg-primary')
    );

    //click on 1st in calendar
    fireEvent.click(getByText('1'));
    const newDate = new Date(initialValue.setDate(1));
    await waitFor(() => {
      expect(getByText(1).classList).toContain('bg-primary');
      expect(container.querySelector('input')?.value).toEqual(
        makeInputValueString(newDate, 'DD/MM/YYYY')
      );
      expect(getByText(day).classList).not.toContain('bg-primary');
    });
  });
  it('initialValue passed in a same date from new Date, should not throw a warning', () => {
    const initialValue = new Date();
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    render(<DatePicker initialValue={initialValue} />);

    expect(consoleSpy).not.toHaveBeenCalled();
    jest.clearAllMocks();
  });
  it('initialValue passed in a different date from new Date, should throw a warning', () => {
    const initialValue = new Date('2020-01-01');
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    render(<DatePicker initialValue={initialValue} />);

    expect(consoleSpy).toHaveBeenCalled();
    jest.clearAllMocks();
  });
  it('initialValue passed with same value as displayDate prop, should not throw a warning', () => {
    const initialValue = new Date('2020-01-01');
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    render(
      <DatePicker initialValue={initialValue} displayDate={initialValue} />
    );

    expect(consoleSpy).not.toHaveBeenCalled();
    jest.clearAllMocks();
  });
  it('when displayDate prop passed, Calendar should reflect correct month', async () => {
    const { getByText, container } = render(
      <DatePicker displayDate={new Date('2020-01-01')} />
    );

    fireEvent.focus(container.querySelector('input')!);

    await waitFor(() => expect(getByText('January 2020')).toBeInTheDocument());
  });

  it('when required pass, should forward the required attr to input', () => {
    const { container, rerender } = render(<DatePicker />);

    expect(container.querySelector('input')).not.toHaveAttribute('required');
    rerender(<DatePicker required />);
    expect(container.querySelector('input')).toHaveAttribute('required');
  });
  it('when className pass, should forward the className  to input', () => {
    const { container, rerender } = render(<DatePicker />);

    expect(container.querySelector('input')?.classList).not.toContain('foo');
    rerender(<DatePicker className="foo" />);
    expect(container.querySelector('input')?.classList).toContain('foo');
  });
  it('when placeholder pass, should forward the placeholder attr  to input', () => {
    const { container, rerender } = render(<DatePicker />);

    expect(container.querySelector('input')).toHaveAttribute(
      'placeholder',
      'dd/mm/yyyy'
    );
    rerender(<DatePicker placeholder="foo" />);
    expect(container.querySelector('input')).toHaveAttribute(
      'placeholder',
      'foo'
    );
  });
  it('when dateFormat changes, placeholder updates accordingly', () => {
    const { container, rerender } = render(
      <DatePicker dateFormat="MM/DD/YYYY" />
    );
    expect(container.querySelector('input')).toHaveAttribute(
      'placeholder',
      'mm/dd/yyyy'
    );

    rerender(<DatePicker dateFormat="YYYY/MM/DD" />);
    expect(container.querySelector('input')).toHaveAttribute(
      'placeholder',
      'yyyy/mm/dd'
    );
  });
  it('when dateFormat changes, value clicked should reflect accordingly', async () => {
    const { container, rerender, getByText } = render(
      <DatePicker dateFormat="MM/DD/YYYY" />
    );

    fireEvent.focus(container.querySelector('input')!);
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument();
      expect(getByText('1')).toBeInTheDocument();
    });

    fireEvent.click(getByText('1'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).not.toBeInTheDocument();
    });
    const newDate = new Date();
    newDate.setDate(1);
    const expectedDate = makeInputValueString(newDate, 'MM/DD/YYYY');
    expect(container.querySelector('input')?.value).toEqual(expectedDate);

    rerender(<DatePicker dateFormat="YYYY/MM/DD" />);
    fireEvent.focus(container.querySelector('input')!);
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument();
      expect(getByText('1')).toBeInTheDocument();
    });

    fireEvent.click(getByText('1'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).not.toBeInTheDocument();
    });
    const newExpectedDate = makeInputValueString(newDate, 'YYYY/MM/DD');
    expect(container.querySelector('input')?.value).toEqual(newExpectedDate);
  });

  it('when disabled prop passed, should forward disabled attribute to button and input', () => {
    const { container } = render(<DatePicker disabled />);

    expect(container.querySelector('input')).toHaveAttribute('disabled');
    expect(container.querySelector('button')).toHaveAttribute('disabled');
  });

  it('autoFocus loads DatePicker with Popover shown', () => {
    const { container } = render(<DatePicker autoFocus />);

    expect(container.querySelector('.popover')).toBeInTheDocument();
  });
  it('onChangeDate fn fires when dates clicked', () => {
    const mockFn = jest.fn();
    const { getByText } = render(
      <DatePicker autoFocus onChangeDate={mockFn} />
    );
    fireEvent.click(getByText('1'));

    expect(mockFn).toHaveBeenCalled();
  });
  it('onFocus and onBlur fn fires when input focus and not focus', () => {
    const focusFn = jest.fn();
    const blurFn = jest.fn();
    const { container } = render(
      <DatePicker onFocus={focusFn} onBlur={blurFn} />
    );
    fireEvent.focus(container.querySelector('input')!);

    expect(focusFn).toHaveBeenCalled();

    fireEvent.blur(container.querySelector('input')!);
    expect(blurFn).toHaveBeenCalled();
    expect(focusFn).toHaveBeenCalledTimes(1);
    expect(blurFn).toHaveBeenCalledTimes(1);
  });
  it('onChange and onClear fn fires when click on Clear button', async () => {
    const onChangeDate = jest.fn();
    const onClear = jest.fn();
    const { container } = render(
      <DatePicker
        onChangeDate={onChangeDate}
        onClear={onClear}
        initialValue={new Date()}
      />
    );
    expect(container.querySelector('input')?.value).toEqual(
      makeInputValueString(new Date(), 'DD/MM/YYYY')
    );

    fireEvent.click(container.querySelector('button')!);

    await waitFor(() => {
      expect(onChangeDate).toHaveBeenCalledTimes(1);
      expect(onClear).toHaveBeenCalledTimes(1);
      expect(container.querySelector('input')?.value).toEqual('');
    });
  });
  it('calendarPlacement forwards its value to the data-popper-placement attr', async () => {
    const { container, rerender } = render(<DatePicker autoFocus />);
    //default should be bottom
    expect(container.querySelector('.popover')).toBeInTheDocument();
    expect(container.querySelector('.popover')?.classList).toContain(
      'bs-popover-bottom'
    );
    await waitFor(() => {
      expect(container.querySelector('.popover')).toHaveAttribute(
        'data-popper-placement',
        'bottom'
      );
    });

    rerender(<DatePicker autoFocus calendarPlacement="top" />);
    await waitFor(() => {
      expect(container.querySelector('.popover')?.classList).toContain(
        'bs-popover-top'
      );
      expect(container.querySelector('.popover')).toHaveAttribute(
        'data-popper-placement',
        'top'
      );
    });
  });
});

describe('Datepicker Range mode', () => {
  afterEach(() => jest.clearAllMocks());
  it('placeholder should show range by default', () => {
    const { container } = render(<DatePicker mode="range" />);
    expect(container.querySelector('input')?.placeholder).toEqual(
      'dd/mm/yyyy - dd/mm/yyyy'
    );
  });
  it('placeholder should reflect dateFormat correctly by default', () => {
    const { container, rerender } = render(
      <DatePicker mode="range" dateFormat="MM/DD/YYYY" />
    );
    expect(container.querySelector('input')?.placeholder).toEqual(
      'mm/dd/yyyy - mm/dd/yyyy'
    );
    rerender(<DatePicker mode="range" dateFormat="YYYY/MM/DD" />);
    expect(container.querySelector('input')?.placeholder).toEqual(
      'yyyy/mm/dd - yyyy/mm/dd'
    );
  });
  it('allows selection dates', async () => {
    const displayDate = new Date('2020-01-01');
    const { getByText, container } = render(
      <DatePicker mode="range" autoFocus displayDate={displayDate} />
    );
    expect(container.querySelector('.popover')).toBeInTheDocument();
    fireEvent.click(getByText('1'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument();
      expect(getByText('1').classList).toContain('bg-primary');
    });

    fireEvent.click(getByText('20'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).not.toBeInTheDocument();
      expect(container.querySelector('input')?.value).toEqual(
        `01/01/2020 - 20/01/2020`
      );
    });
    fireEvent.focus(container.querySelector('input')!);
    const oneToTwenty = Array.from({ length: 20 }, (_, i) => i + 1);
    await waitFor(() => {
      oneToTwenty.forEach((day) => {
        expect(getByText(day).classList).toContain('bg-primary');
      });
      expect(getByText('21').classList).not.toContain('bg-primary');
    });
  });

  it('allows selection date, click of next icon , selection of second date', async () => {
    const displayDate = new Date('2020-01-01');
    const { getByText, container } = render(
      <DatePicker mode="range" autoFocus displayDate={displayDate} />
    );
    fireEvent.click(getByText('28'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument();
      expect(getByText('28').classList).toContain('bg-primary');
    });

    fireEvent.click(container.querySelector('i.bi-chevron-right')!);
    await waitFor(() => {
      expect(getByText('February 2020')).toBeInTheDocument();
    });
    fireEvent.click(getByText('2'));
    await waitFor(() => {
      expect(container.querySelector('.popover')).not.toBeInTheDocument();
      expect(container.querySelector('input')?.value).toEqual(
        `28/01/2020 - 02/02/2020`
      );
    });
    fireEvent.focus(container.querySelector('input')!);
    await waitFor(() => {
      [1, 2].forEach((day) => {
        expect(getByText(day).classList).toContain('bg-primary');
      });
      expect(getByText('3').classList).not.toContain('bg-primary');
    });
    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    await waitFor(() => {
      expect(getByText('January 2020')).toBeInTheDocument();
      [28, 29, 30, 31].forEach((day) => {
        expect(getByText(day).classList).toContain('bg-primary');
      });
      expect(getByText('27').classList).not.toContain('bg-primary');
    });
  });

  it('when initialValue passed in, either start or end should have same value as displayDate, else a console.error should appear', async () => {
    const initialValue = { start: new Date('2020-01-01'), end: new Date() };
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    const { rerender } = render(
      <DatePicker mode="range" initialValue={initialValue} />
    );
    expect(consoleSpy).not.toHaveBeenCalled();

    const newInitialValue = {
      start: new Date('2022-12-12'),
      end: new Date('2022-12-14'),
    };
    rerender(
      <DatePicker
        mode="range"
        initialValue={newInitialValue}
        displayDate={new Date('2022-12-12')}
      />
    );

    expect(consoleSpy).not.toHaveBeenCalled();

    rerender(
      <DatePicker
        mode="range"
        initialValue={newInitialValue}
        displayDate={new Date('2022-12-20')}
      />
    );
    expect(consoleSpy).toHaveBeenCalled();
    expect(consoleSpy).toHaveBeenCalledTimes(1);
  });

  it('when range mode, initialValue should be RangeSelectionValue type , else console error', () => {
    const initialValue = new Date();
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    const { rerender } = render(
      <DatePicker
        mode="range"
        initialValue={initialValue}
        displayDate={initialValue}
      />
    );
    expect(consoleSpy).toHaveBeenCalled();
    jest.clearAllMocks();
    const newValue = { start: new Date(), end: new Date() };
    rerender(
      <DatePicker
        mode="range"
        initialValue={newValue}
        displayDate={new Date()}
      />
    );
    expect(consoleSpy).not.toHaveBeenCalled();
  });
  it('in range mode, start date must be ealrier than end date , else throw warning', () => {
    const initialValue = {
      start: new Date('2020-01-25'),
      end: new Date('2020-01-23'),
    };
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    const { rerender } = render(
      <DatePicker
        mode="range"
        initialValue={initialValue}
        displayDate={new Date('2020-01-25')}
      />
    );
    expect(consoleSpy).toHaveBeenCalled();
    jest.clearAllMocks();

    rerender(
      <DatePicker
        mode="range"
        initialValue={{
          start: new Date('2020-01-23'),
          end: new Date('2020-01-25'),
        }}
        displayDate={new Date('2020-01-25')}
      />
    );
    expect(consoleSpy).not.toHaveBeenCalled();
  });

  it('allows selection of second date that is ealrier than first selected date, input should fix accordingly', async() => {
    const { getByText, container } = render(
      <DatePicker
        mode="range"
        autoFocus
        displayDate={new Date("2020-01-01")}
      />
    );
    fireEvent.click(getByText('20'))

    await waitFor(() => {
      expect(getByText('20').classList).toContain('bg-primary')
      expect(container.querySelector('input')?.value).toEqual(`20/01/2020 - ` )
    })
    fireEvent.click(getByText('15'))
    await waitFor(() => {
      expect(container.querySelector('input')?.value).toEqual(`15/01/2020 - 20/01/2020` )
    })
    //new selection
    fireEvent.focus(container.querySelector('input')!)
    await waitFor(() => {
      expect(container.querySelector('.popover')).toBeInTheDocument()
      const selecteDates = [15,16,17,18,19,20]
      selecteDates.forEach(day => expect(getByText(day).classList).toContain('bg-primary'))
    })

    fireEvent.click(getByText('1'))
    await waitFor(() => expect(container.querySelector('input')?.value).toEqual('01/01/2020 - '))
    fireEvent.click(getByText('20'))
    await waitFor(() => expect(container.querySelector('input')?.value).toEqual('01/01/2020 - 20/01/2020'))

  })
});
