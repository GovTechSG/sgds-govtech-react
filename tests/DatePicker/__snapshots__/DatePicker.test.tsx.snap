// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker has the default html structure 1`] = `
<DocumentFragment>
  <div
    class="input-group input-group-md sgds"
    variant="has-icon"
  >
    <div>
       
      <input
        class="form-control"
        placeholder="dd/mm/yyyy"
        readonly=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary sgds"
      type="button"
    >
      <i
        class="bi bi-x"
      />
    </button>
    <i
      class="bi bi-calendar form-control-icon"
    />
  </div>
</DocumentFragment>
`;
